text/1029-rbac.rst

```diff
@@ -0,0 +1,340 @@
+It is evaluated to a boolean value, which is true iff the role of current
+connection has the associated permission.
+
+The global cannot be assigned to, but it is otherwise a normal global variable;
+it can be used in any expression.
+
+In combination with access policies, this allows precise control over which
+objects of an object type are visible or can be modified by each role.
+
+For example::
+
+    CREATE TYPE Secret {
+        CREATE PROPERTY super_secret: str;
+
+        CREATE ACCESS POLICY allow_webapp
+            ALLOW ALL
+            USING (global frontend::webapp);
+
+        CREATE ACCESS POLICY allow_warehouse
+            ALLOW SELECT
+            USING (global data_export);
+
+        CREATE ACCESS POLICY allow_inserts
+            ALLOW INSERT;
+    };
+
+An important detail is that access policies for ``select`` only filter data
+and do not raise errors when an object type is accessed without required
+permissions. It would be more user-friendly to raise an error instead of
+returning no results, but we currently do not provide a reliable way of
+raising errors within access policies. See rejected ideas for justification.
+
+Per-function access control
+---------------------------
+
+Functions cannot have access policies, but we do want to be able to
+limit access to certain functions, such as ``sys::reset_query_stats`` and
+``net::http::schedule_request``.
+
+To provide access control to functions, we introduce a new function metadata
+field named ``require_permission``. It adds the provided permission to the list
+of required permissions of the query, similar to how `DDL` and other existing
+capabilities work.
+
+code::
+
+    CREATE FUNCTION sys::reset_query_stats(...) -> ... {
+        SET volatility := ...
+        SET require_permission := sys::query_stats;
+        USING SQL FUNCTION ...;
+    }; 
+
+When a query is compiled the set of required permissions is stored on the
+compiled query unit. When executing the unit, the server validates that all
+permissions of this set are held by the role of current connection and that
+these permissions are allowed by the protocol capabilities allowed by the
+client.
+
+
+Future work
+===========
+
+Converting more of SUPERUSER to permissions
+-------------------------------------------
+
+With current design non-SUPERUSER rols have a very limited access to the
+database. Namely, they cannot apply migrations, use in-language transaction
+commands, modify or view config, use explain, describe or administer.
+
+We will probably want to convert access to these commands into permissions
+defined in ``sys`` module, similar to ``sys::data_modification``.
+
+Before this RFC, non-SUPERUSER roles has practically no permissions. This RFC
+aims to grant as little access as possible, while still being practical.
+That's because granting new access is a non-breaking change, but taking away
+access is a breaking change.
+
+
+Permission inheritance
+----------------------
```

Comment:
It's possible we might also want to organize this in the opposite direction, with `permission groups`, or something?

Reply:
Ah, roles have inheritance, though!